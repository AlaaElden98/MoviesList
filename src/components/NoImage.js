import React from 'react';
import {Text, View, StyleSheet} from 'react-native';

export function NoImage() {
  return (
    <View style={styles.container}>
      <Text>Image not availiable</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 10,
    borderWidth: 2,
  },
});
