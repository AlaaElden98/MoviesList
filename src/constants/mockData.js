export const movies = [
  {
    poster_path: '/nkayOAUBUu4mMvyNf9iHSUiPjF1.jpg',
    id: 578701,
    overview: 'Mock overview 1',
    title: 'Mock title 1',
  },
  {
    poster_path: '/iDdpiUnCeXvBmrkBFpL6lKsZt33.jpg',
    id: 578702,
    overview: 'Mock overview 1',
    title: 'Mock title 1',
  },
];

export const movie = {
  poster_path: '/mockPath' || null,
  title: 'mockTitle',
  overview: 'mockOverView',
  date: '123/123/123',
};
