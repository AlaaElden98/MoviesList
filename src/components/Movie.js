import React from 'react';
import {Text, View, Image, StyleSheet} from 'react-native';

import {NoImage} from './NoImage';
import {responsiveFontSize, responsiveWidth} from '../utils/helperFunctions';

export function Movie(props) {
  const {movie, imageBaseUrl} = props;
  return (
    <View
      style={styles.container}
      accessibilityLabel={
        'Movie item includes poster,title,overview,and the release date of the movie'
      }>
      {imageBaseUrl ? (
        <Image
          source={{
            uri: imageBaseUrl + 'w500' + movie.poster_path,
          }}
          // Test this
          style={{
            width: responsiveWidth(30),
            height: responsiveWidth(65),
            borderRadius: 8,
          }}
          resizeMode="stretch"
          accessible={true}
          accessibilityLabel={`Poster of ${movie.title} movie`}
        />
      ) : (
        <NoImage />
      )}

      <View style={styles.textDataContainer}>
        <View style={styles.titleContainer}>
          <Text
            numberOfLines={1}
            ellipsizeMode="tail"
            style={styles.title}
            accessibilityLabel={`Movie name: ${movie.title}`}>
            {movie.title}
          </Text>
        </View>
        <View style={styles.overViewContainer}>
          <Text
            numberOfLines={8}
            ellipsizeMode="tail"
            style={styles.overview}
            accessibilityLabel={`Movie overview: ${movie.overview}`}>
            {movie.overview}
          </Text>
        </View>
        <View style={styles.releaseDateContainer}>
          <Text
            style={styles.releaseDate}
            accessibilityLabel={`Release date : ${movie.release_date}`}>
            Release date : {movie.release_date}
          </Text>
        </View>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
    padding: 10,
    margin: 8,
    backgroundColor: 'white',
    borderRadius: 20,
    borderWidth: 0,
  },
  textDataContainer: {
    flex: 1,
  },
  overview: {fontSize: responsiveFontSize(2)},
  overViewContainer: {
    flex: 1,
    margin: 5,
    padding: 5,
    alignItems: 'center',
  },
  title: {
    fontSize: responsiveFontSize(3),
    fontWeight: 'bold',
    textDecorationLine: 'underline',
  },
  titleContainer: {
    alignItems: 'center',
    margin: 5,
  },
  releaseDateContainer: {
    flex: 1,
    alignItems: 'flex-start',
    justifyContent: 'flex-end',
    marginLeft: 5,
    padding: 5,
  },
  releaseDate: {
    fontSize: responsiveFontSize(2),
    color: 'grey',
    fontStyle: 'italic',
  },
});
