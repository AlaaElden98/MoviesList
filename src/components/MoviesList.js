import React, {useEffect, useState} from 'react';
import {FlatList} from 'react-native';

import {Movie} from './Movie';
import {getMoviesList} from '../api/api';
import {CustomActivityIndicator} from './CustomActivityIndicator';

export function MoviesList(props) {
  const [page, setPage] = useState(0);
  const [totalPages, setTotalPages] = useState(0);
  const [movies, setMovies] = useState([]);
  const {imageBaseUrl} = props;

  const getMoviesData = async pageNumber => {
    const data = await getMoviesList(pageNumber);
    movies ? setMovies([...movies, ...data.results]) : setMovies(data.results);
    setPage(data.page);
    setTotalPages(data.total_pages);
  };

  useEffect(() => {
    getMoviesData();
  }, []);

  const onEndReached = () => {
    getMoviesData(page < totalPages ? page + 1 : page);
  };

  const renderItem = ({item}) => {
    return <Movie movie={item} imageBaseUrl={imageBaseUrl} />;
  };

  return (
    <FlatList
      data={movies}
      renderItem={renderItem}
      keyExtractor={item => item.id}
      onEndReachedThreshold={0.5}
      onEndReached={onEndReached}
      ListFooterComponent={CustomActivityIndicator}
    />
  );
}
