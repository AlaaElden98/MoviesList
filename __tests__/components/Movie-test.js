import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import {Movie} from '../../src/components/Movie';
import {movie} from '../../src/constants/mockData';

test('Movie renders correctly', () => {
  const tree = renderer.create(<Movie movie={movie} />).toJSON();
  expect(tree).toMatchSnapshot();
});
