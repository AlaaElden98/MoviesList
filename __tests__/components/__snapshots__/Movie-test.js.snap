// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders movie item 1`] = `
<View
  accessibilityLabel="Movie item includes poster,title,overview,and the release date of the movie"
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 20,
      "borderWidth": 0,
      "flex": 1,
      "flexDirection": "row",
      "margin": 8,
      "padding": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderRadius": 10,
        "borderWidth": 2,
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessibilityLabel="Image not availiable"
    >
      Image not availiable
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "margin": 5,
        }
      }
    >
      <Text
        accessibilityLabel="Movie name: mockTitle"
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Object {
            "fontSize": 45.89389937671454,
            "fontWeight": "bold",
            "textDecorationLine": "underline",
          }
        }
      >
        mockTitle
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "margin": 5,
          "padding": 5,
        }
      }
    >
      <Text
        accessibilityLabel="Movie overview: mockOverView"
        ellipsizeMode="tail"
        numberOfLines={8}
        style={
          Object {
            "fontSize": 30.595932917809698,
          }
        }
      >
        mockOverView
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flex": 1,
          "justifyContent": "flex-end",
          "marginLeft": 5,
          "padding": 5,
        }
      }
    >
      <Text
        accessibilityLabel="Release date : undefined"
        style={
          Object {
            "color": "grey",
            "fontSize": 30.595932917809698,
            "fontStyle": "italic",
          }
        }
      >
        Release date : 
      </Text>
    </View>
  </View>
</View>
`;
